### PR TITLE
Add `move_angle_toward` function

### PR DESCRIPTION
**Pull request by antonWetzel**
[Proposal #4194](https://github.com/godotengine/godot-proposals/issues/4194#issue-1164233630)

Add `move_toward_angle` function
  - like `move_toward`, but direction regards wrap around `TAU`
  - added to `@GlobalScope` for GDScript
  - added to `Mathf` for CSharp
  - added as build-in function to VisualScript
  - add unit test for c-version

### CODE REVIEW

**Review by aaronfranke — CHANGES_REQUESTED** [review]
The documentation should be updated to note that all parameters are angles in radians.

We should include test cases for this method in the C++ code to ensure it works correctly.

I remember someone mentioned the name `move_toward_angle`. I'm not sure which name is preferred, I think both are fine, I'm just putting it up for discussion. Does anyone has any strong feelings one way or the other?

See also:

* Merged lerp_angle method #30693
* Proposed angle_diff method #37078

**Comment by antonWetzel** [comment]
> We should include test cases for this method in the C++ code to ensure it works correctly.

I could not find tests for any function in `core/math/math_funcs.h`; should I create `tests/core/test_math_funcs.h` or add the test somewhere else?
- draft for `tests/core/test_math_funcs.h` is below
- function could be tested as expressions in GDScript

> I remember someone mentioned the name `move_toward_angle`. I'm not sure which name is preferred, I think both are fine, I'm just putting it up for discussion. Does anyone has any strong feelings one way or the other?

`move_angle_toward_angle` is easy to understand, but too long. I have no preference which `angle` should be omitted, but because the function is a variant of `move_toward` I think i will change it to `move_toward_angle`.

### draft for `tests/core/test_math_funcs.h`
```c
[...Lisence...]

#ifndef TEST_MATH_FUNCS_H
#define TEST_MATH_FUNCS_H

#include "core/math/math_funcs.h"
#include "core/math/math_defs.h"
#include "tests/test_macros.h"

namespace TestMathFuncs {
TEST_CASE("[MathFuncs] Angle methods") {
	CHECK_MESSAGE(
			Math::is_equal_approx(Math::move_angle_toward(0.1, Math_TAU - 0.2, 0.2), -0.1), 
			"MathFuncs move_angle_toward should wrap as expected.");
	CHECK_MESSAGE(
			Math::is_equal_approx(Math::move_angle_toward(0.1, 0.2, 0.3), 0.2), 
			"MathFuncs move_angle_toward should stop as expected.");
}
} // namespace TestMathFuncs

#endif // TEST_MATH_FUNCS_H

```

**Comment by aaronfranke** [comment]
@antonWetzel Adding a new file for this sounds good. Indeed we are currently missing tests for lots of math functions, we should add them eventually. From looking at the folder, one might assume that the `test_math` file tests math functions, but it actually tests a bunch of things and should probably be deleted eventually. So we should put math function tests in `test_math_funcs`.

**Comment by antonWetzel** [comment]
@aaronfranke I changed the name and added a test some time ago.
Should I do somehting to continue to pull request?

File: modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml [hunk 1]
```diff
@@ -105,116 +105,119 @@
 		<constant name="MATH_MOVE_TOWARD" value="30" enum="BuiltinFunc">
 			Moves the number toward a value, based on the third input.
 		</constant>
-		<constant name="MATH_RANDOMIZE" value="31" enum="BuiltinFunc">
+		<constant name="MATH_MOVE_TOWARD_ANGLE" value="31" enum="BuiltinFunc">
+			Moves the number toward a angle, based on the third input.
```
Comment (aaronfranke) [inline modules/visual_script/doc_classes/VisualScriptBuiltinFunc.xml:109]:
```suggestion
			Moves the first angle toward the second angle, based on the third input.
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs [hunk 2]
```diff
@@ -445,6 +445,28 @@ public static real_t MoveToward(real_t from, real_t to, real_t delta)
             return from + (Sign(to - from) * delta);
         }
 
+        /// <summary>
+        /// Moves <paramref name="from"/> toward <paramref name="to"/> by the <paramref name="delta"/> angle (in radians).
```
Comment (aaronfranke) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:449]: This line looks a bit long, we can split it up:

```suggestion
        /// Moves <paramref name="from"/> toward <paramref name="to"/>
        /// by the <paramref name="delta"/> angle (in radians).
```

**Review by aaronfranke — COMMENTED** [review]
This PR looks mostly good. I left some documentation suggestions below, and this needs [a squash](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html), and then I'd like to manually test it before giving it a ✔️,

**Comment by antonWetzel** [comment]
All commits should be squashed into one.

**Comment by aaronfranke** [comment]
@antonWetzel The code doesn't behave as expected. I think you need more test cases.

```
print(move_toward_angle(TAU - 0.2, 0.2, 0.3))
```

I would expect this code to return `0.1` (or approximately that), but instead it returns `6.38318530717959`.

**Comment by antonWetzel** [comment]
@aaronfranke The result is TAU + 0.1
For a Node2D the rotation is not clamped to [-PI/PI], therefore I thougt rotation must not be clamped.
The current implementation has a the result close to the input:  abs(result - input) <= abs(delta).
I have no preference if the result should be clamped or close to the input.

**Comment by antonWetzel** [comment]
Self closed because of inactivity.

If someone needs this they can use
```gdscript
#mathp.gd
class_name Mathp

static func normalize_angle(angle: float) -> float:
	return angle - round(angle / TAU) * TAU

static func move_toward_angle(from: float, to: float, delta: float):
	var diff := normalize_angle(to - from)
	var s := sign(diff) as float
	return normalize_angle(from + min(delta, diff * s) * s)

```
or
```csharp
//Mathp.cs
using System;
using Godot;

static class Mathp {

	public static float NormalizeAngle(float angle) {
		return angle - Mathf.Round(angle / Mathf.Tau) * Mathf.Tau;
	}

	public static float MoveTowardAngle(float from, float to, float delta) {
		float diff = NormalizeAngle(to - from);
		float sign = Mathf.Sign(diff);
		return NormalizeAngle(from + Mathf.Min(delta, diff * sign) * sign);
	}
}

```

**Comment by YuriSizov** [comment]
> Self closed because of inactivity.

FYI this PR has been assessed as okay to include, but not critical for 4.0 (so it can be added in later 4.x versions).

**Pull request closed without merging**